Switch hardened derivation marker to h

Sjors opened the pull request:
This makes it easier to handle descriptor strings manually, especially when importing from another Bitcoin Core wallet.

For example the `importdescriptors` RPC call is easiest to use `h` as the marker: `'["desc": ".../0h/..."]'`, avoiding the need for escape characters. With this change `listdescriptors` will use `h`, so you can copy-paste the result, without having to add escape characters or switch `'` to 'h' manually.

Both markers can still be parsed.

The `hdkeypath` field in `getaddressinfo` is also impacted by this change, except for legacy wallets. The latter is to prevent accidentally breaking ancient software that uses our legacy wallet. 

See discussion in #15740

furszy reviewed (commented):
Concept ACK

luke-jr commented on doc/release-notes-26076.md:
```diff
@@ -0,0 +1,9 @@
+
+- The `deriveaddresses`, `getdescriptorinfo` and `listdescriptors` RPC's now
+  use `h` rather than apostrophe (`'`) to indicate hardened derivation. This
+  makes it easier to handle descriptor strings manually. E.g. an RPC call that
```
The enclosing single quotes are shell escaping, not part of the actual RPC call (which has no issue at present).

Sjors replied:
By "manually" I mostly mean using `bitcoin-cli` from the shell.

jarolrod commented:
concept ack

achow101 commented:
ACK fc5594d036a47434ea301f5e6adff06c59649fd2

Sjors commented:
Made a slight documentation change to the first commit and then pushed three more which change the following behaviour:

1. Normalized descriptors use `h` (this is useful when using `listwalletdescriptors` with older wallets)
2. Non-normalized descriptors (including the private string) use whathever was provided at creation time, as @sipa requested
3. New wallets use `h` instead of `\'`

I could squash the new behavior into the original commit, but I'll wait for concept ACK.

achow101 commented:
Concept ACK

Please squash

Sjors commented:
@achow101 squashed

achow101 commented on src/test/descriptor_tests.cpp:
```diff
@@ -137,16 +137,14 @@ void DoCheck(const std::string& prv, const std::string& pub, const std::string&
+        prv = UseHInsteadOfApostrophe(prv);
     }
+    parse_priv = Parse(prv, keys_priv, error);
     BOOST_CHECK_MESSAGE(parse_priv, error);
     if (replace_apostrophe_with_h_in_pub) {
-        parse_pub = Parse(UseHInsteadOfApostrophe(pub), keys_pub, error);
-    } else {
-        parse_pub = Parse(pub, keys_pub, error);
+        pub = UseHInsteadOfApostrophe(pub);
     }
+    parse_pub = Parse(pub, keys_pub, error);
     BOOST_CHECK_MESSAGE(parse_pub, error);
```
It's not clear to me why this and the change in `Check` are necessary?

Sjors replied:
Could be a left-over from the squash. Will look into this once we decide whether or not to normalise private keys, because that might change things again: https://github.com/bitcoin/bitcoin/pull/26076#discussion_r1016668800

Sjors replied:
Not a left-over; the test fails without it. The behaviour is quite different now that we echo the original descriptor rather than always put `'` in it.

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -414,31 +419,35 @@ class BIP32PubkeyProvider final : public PubkeyProvider
         CExtKey key;
         if (!GetExtKey(arg, key)) return false;
-        out = EncodeExtKey(key) + FormatHDKeypath(m_path);
+        out = EncodeExtKey(key) + FormatHDKeypath(m_path, /*apostrophe=*/m_apostrophe);
```
I intentionally didn't normalize the private-key versions. Though I could.

darosior replied:
What was the rationale for only normalizing the public-key version with `h`?

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -1047,14 +1056,18 @@ enum class ParseScriptContext {
 
 /** Parse a key path, being passed a split list of elements (the first element is ignored). */
-[[nodiscard]] bool ParseKeyPath(const std::vector<Span<const char>>& split, KeyPath& out, std::string& error)
+[[nodiscard]] bool ParseKeyPath(const std::vector<Span<const char>>& split, KeyPath& out, std::optional<bool>& apostrophe, std::string& error)
```
In one of the call sites it first checks the origin and then the rest of the path. The optional is necessary to combine those results, especially in the case where the origin doesn't have any hardened derivation (one of the test cases).

achow101 replied:
I'm not seeing how it is necessary. It could initialized to `false` in `ParsePubKey` and I think that would achieve the same result. Nothing is checking whether the optional has a value, except at the very end with `value_or(false)`, and that's equivalent to initializing as `false`.

Sjors replied:
If I initialise to `false`, it would be set to `true` when checking the origin and then back to `false` when checking the path.

achow101 replied:
This seems to work:

```diff
diff --git a/src/script/descriptor.cpp b/src/script/descriptor.cpp
index e639e494582..877d2538b2c 100644
--- a/src/script/descriptor.cpp
+++ b/src/script/descriptor.cpp
@@ -1056,7 +1056,7 @@ enum class ParseScriptContext {
 };
 
 /** Parse a key path, being passed a split list of elements (the first element is ignored). */
-[[nodiscard]] bool ParseKeyPath(const std::vector<Span<const char>>& split, KeyPath& out, std::optional<bool>& apostrophe, std::string& error)
+[[nodiscard]] bool ParseKeyPath(const std::vector<Span<const char>>& split, KeyPath& out, bool& apostrophe, std::string& error)
 {
     for (size_t i = 1; i < split.size(); ++i) {
         Span<const char> elem = split[i];
@@ -1083,7 +1083,7 @@ enum class ParseScriptContext {
 }
 
 /** Parse a public key that excludes origin information. */
-std::unique_ptr<PubkeyProvider> ParsePubkeyInner(uint32_t key_exp_index, const Span<const char>& sp, ParseScriptContext ctx, FlatSigningProvider& out, std::optional<bool>& apostrophe, std::string& error)
+std::unique_ptr<PubkeyProvider> ParsePubkeyInner(uint32_t key_exp_index, const Span<const char>& sp, ParseScriptContext ctx, FlatSigningProvider& out, bool& apostrophe, std::string& error)
 {
     using namespace spanparsing;
 
@@ -1149,7 +1149,7 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(uint32_t key_exp_index, const S
         extpubkey = extkey.Neuter();
         out.keys.emplace(extpubkey.pubkey.GetID(), extkey.key);
     }
-    return std::make_unique<BIP32PubkeyProvider>(key_exp_index, extpubkey, std::move(path), type, apostrophe.value_or(false));
+    return std::make_unique<BIP32PubkeyProvider>(key_exp_index, extpubkey, std::move(path), type, apostrophe);
 }
 
 /** Parse a public key including origin information (if enabled). */
@@ -1163,7 +1163,7 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(uint32_t key_exp_index, const Span<c
         return nullptr;
     }
     // This is set if either the origin or path suffix contains a hardened deriviation.
-    std::optional<bool> apostrophe;
+    bool apostrophe = false;
     if (origin_split.size() == 1) {
         return ParsePubkeyInner(key_exp_index, origin_split[0], ctx, out, apostrophe, error);
     }
@@ -1190,7 +1190,7 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(uint32_t key_exp_index, const Span<c
     if (!ParseKeyPath(slash_split, info.path, apostrophe, error)) return nullptr;
     auto provider = ParsePubkeyInner(key_exp_index, origin_split[1], ctx, out, apostrophe, error);
     if (!provider) return nullptr;
-    return std::make_unique<OriginPubkeyProvider>(key_exp_index, std::move(info), std::move(provider), apostrophe.value_or(false));
+    return std::make_unique<OriginPubkeyProvider>(key_exp_index, std::move(info), std::move(provider), apostrophe);
 }
 
 std::unique_ptr<PubkeyProvider> InferPubkey(const CPubKey& pubkey, ParseScriptContext, const SigningProvider& provider)

```

Sjors replied:
Ah yes, because `ParseKeyPath` leaves `&apostrophe` alone if it doesn't find hardened derivation. Fixed and added some documentation.

JeremyRubin commented:
it might make sense to have new wallets only generate `h` by default if opted into via flag since there is still lots of software that might not be compatible with a new format.

Sjors commented:
@JeremyRubin is there software that can't parse BIP32 paths with `'`? I'd rather stalk them a bit during the next six months (before v25.0) than work around the issue :-)

JeremyRubin commented:
ah nope i misremembered that rust-bitcoin didn't support

achow101 commented:
ACK de83b8de1a3fee4134ceccadd0d09f5f8a5d3460

sipa reviewed (commented):
utACK de83b8de1a3fee4134ceccadd0d09f5f8a5d3460

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2291,7 +2291,7 @@ bool DescriptorScriptPubKeyMan::SetupDescriptorGeneration(const CExtKey& master_
-        desc_prefix += "/1'";
+        desc_prefix += "/1h";
     } else {
         desc_prefix += "/0'";
```
Missed a `'` here and one below at line 2300.

Sjors replied:
Thanks, I also updated the `combo()` descriptor in `MigrateToDescriptor`.

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/26076/commits/e4a5fa9312920737b10ce9439ada353f6c78340d

Sjors commented:
> There's a `FormatHDKeypath` in 'LegacyScriptPubKeyMan::MigrateToDescriptor`that should set`apostrophe=false`

Fixed.

furszy commented on src/script/descriptor.cpp:
```diff
@@ -1148,7 +1170,11 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(uint32_t key_exp_index, const Span<c
         return nullptr;
     }
-    if (origin_split.size() == 1) return ParsePubkeyInner(key_exp_index, origin_split[0], ctx, out, error);
+    // This is set if either the origin or path suffix contains a hardened deriviation.
```
s/deriviation/derivation

furszy commented on src/script/descriptor.cpp:
```diff
@@ -1046,15 +1055,27 @@ enum class ParseScriptContext {
+/**
+ * Parse a key path, being passed a split list of elements (the first element is ignored).
+ *
+ * @param[in] split BIP32 path string, using either ' or h for hardened deriviation
```
s/deriviation/derivation

furszy reviewed (commented):
Light ACK 6850bc55

not blocking point:
To not mixup flows, would had liked to see two commits instead of one. The first one with the apostrophe read changes (the `m_apostrophe` field set flow) and a second one only containing the behavior change (the retrieval of descriptors using ‘h’ instead of the apostrophe).

furszy reviewed (commented):
Also, have you thought about moving the `getaddressinfo` "hdkeypath" return value to use 'h' instead of the apostrophe?
Atm, the value uses the legacy `WriteHDKeypath` function, which always sets `apostrophe=true`.

achow101 commented:
ACK 6850bc550c67a55a939f0a47690badcc4d3b4b25

darosior commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1822,7 +1822,7 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
             // Make the combo descriptor
             std::string xpub = EncodeExtPubKey(master_key.Neuter());
-            std::string desc_str = "combo(" + xpub + "/0'/" + ToString(i) + "'/*')";
+            std::string desc_str = "combo(" + xpub + "/0h/" + ToString(i) + "h/*')";
```
You are missing one occurrence in this expression:
```suggestion
            std::string desc_str = "combo(" + xpub + "/0h/" + ToString(i) + "h/*h)";
```

darosior commented on src/script/descriptor.cpp:
```diff
@@ -414,31 +419,35 @@ class BIP32PubkeyProvider final : public PubkeyProvider
-    std::string ToString() const override
+    std::string ToString(bool normalized) const
     {
-        std::string ret = EncodeExtPubKey(m_root_extkey) + FormatHDKeypath(m_path);
+        std::string ret = EncodeExtPubKey(m_root_extkey) + FormatHDKeypath(m_path, /*apostrophe=*/m_apostrophe);
         if (IsRange()) {
             ret += "/*";
-            if (m_derive == DeriveType::HARDENED) ret += '\'';
+            if (m_derive == DeriveType::HARDENED) ret += (!normalized && m_apostrophe) ? '\'' : 'h';
```
Looks like it should use `h` everywhere when normalized? (Not only in the hardened-wildcard step.)
```patch
diff --git a/src/script/descriptor.cpp b/src/script/descriptor.cpp
index e3a20c17f4..a2372d7409 100644
--- a/src/script/descriptor.cpp
+++ b/src/script/descriptor.cpp
@@ -421,10 +421,11 @@ public:
     }
     std::string ToString(bool normalized) const
     {
-        std::string ret = EncodeExtPubKey(m_root_extkey) + FormatHDKeypath(m_path, /*apostrophe=*/m_apostrophe);
+        const bool use_apostrophe = !normalized && m_apostrophe;
+        std::string ret = EncodeExtPubKey(m_root_extkey) + FormatHDKeypath(m_path, /*apostrophe=*/use_apostrophe);
         if (IsRange()) {
             ret += "/*";
-            if (m_derive == DeriveType::HARDENED) ret += (!normalized && m_apostrophe) ? '\'' : 'h';
+            if (m_derive == DeriveType::HARDENED) ret += use_apostrophe ? '\'' : 'h';
         }
         return ret;
     }
```

darosior commented on src/script/descriptor.cpp:
```diff
@@ -1046,15 +1055,27 @@ enum class ParseScriptContext {
+            if (last == '\'' || last == 'h') {
+                elem = elem.first(elem.size() - 1);
+                hardened = true;
+                apostrophe = last == '\'';
```
Should we keep the current behaviour of serializing with apostrophes if *any* derivation index contains an apostrophe? I guess it doesn't really matter.

Sjors replied:
I missed that, changing to `h`.

Sjors replied:
Not sure. I think I'd rather have an error message for mixed use of `h` and `'`.

darosior commented on src/test/descriptor_tests.cpp:
```diff
@@ -451,14 +434,14 @@ BOOST_AUTO_TEST_CASE(descriptor_test)
-    CheckUnparsable("sh(multi(2,[00000000/111'/222]xprvA1RpRA33e1JQ7ifknakTFpgNXPmW2YvmhqLQYMmrj4xJXXWYpDPS3xz7iAxn8L39njGVyuoseXzU6rcxFLJ8HFsTjSyQbLYnMpCqE2VbFWc,xprv9uPDJpEQgRQfDcW7BkF7eTya6RPxXeJCqCJGHuCJ4GiRVLzkTXBAJMu2qaMWPrS7AANYqdq6vcBcBUdJCVVFceUvJFjaPdGZ2y9WACViL4L/0))#ggrsrxf", "sh(multi(2,[00000000/111'/222]xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWkY ...
+    CheckUnparsable("sh(multi(2,[00000000/111'/222]xprvA1RpRA33e1JQ7ifknakTFpgNXPmW2YvmhqLQYMmrj4xJXXWYpDPS3xz7iAxn8L39njGVyuoseXzU6rcxFLJ8HFsTjSyQbLYnMpCqE2VbFWc,xprv9uPDJpEQgRQfDcW7BkF7eTya6RPxXeJCqCJGHuCJ4GiRVLzkTXBAJMu2qaMWPrS7AANYqdq6vcBcBUdJCVVFceUvJFjaPdGZ2y9WACViL4L/0))#ggrsrxfyq", "sh(multi(2,[00000000/111'/222]xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHW ...
+    CheckUnparsable("sh(multi(2,[00000000/111'/222]xprvA1RpRA33e1JQ7ifknakTFpgNXPmW2YvmhqLQYMmrj4xJXXWYpDPS3xz7iAxn8L39njGVyuoseXzU6rcxFLJ8HFsTjSyQbLYnMpCqE2VbFWc,xprv9uPDJpEQgRQfDcW7BkF7eTya6RPxXeJCqCJGHuCJ4GiRVLzkTXBAJMu2qaMWPrS7AANYqdq6vcBcBUdJCVVFceUvJFjaPdGZ2y9WACViL4L/0))#ggrsrxf", "sh(multi(2,[00000000/111'/222]xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWkY ...
     CheckUnparsable("sh(multi(3,[00000000/111'/222]xprvA1RpRA33e1JQ7ifknakTFpgNXPmW2YvmhqLQYMmrj4xJXXWYpDPS3xz7iAxn8L39njGVyuoseXzU6rcxFLJ8HFsTjSyQbLYnMpCqE2VbFWc,xprv9uPDJpEQgRQfDcW7BkF7eTya6RPxXeJCqCJGHuCJ4GiRVLzkTXBAJMu2qaMWPrS7AANYqdq6vcBcBUdJCVVFceUvJFjaPdGZ2y9WACViL4L/0))#ggrsrxfy", "sh(multi(3,[00000000/111'/222]xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWk ...
-    CheckUnparsable("sh(multi(2,[00000000/111'/222]xprvA1RpRA33e1JQ7ifknakTFpgNXPmW2YvmhqLQYMmrj4xJXXWYpDPS3xz7iAxn8L39njGVyuoseXzU6rcxFLJ8HFsTjSyQbLYnMpCqE2VbFWc,xprv9uPDJpEQgRQfDcW7BkF7eTya6RPxXeJCqCJGHuCJ4GiRVLzkTXBAJMu2qaMWPrS7AANYqdq6vcBcBUdJCVVFceUvJFjaPdGZ2y9WACViL4L/0))#ggssrxfy", "sh(multi(2,[00000000/111'/222]xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWk ...
-    CheckUnparsable("sh(multi(2,[00000000/111'/222]xprvA1RpRA33e1JQ7ifknakTFpgNXPmW2YvmhqLQYMmrj4xJXXWYpDPS3xz7iAxn8L39njGVyuoseXzU6rcxFLJ8HFsTjSyQbLYnMpCqE2VbFWc,xprv9uPDJpEQgRQfDcW7BkF7eTya6RPxXeJCqCJGHuCJ4GiRVLzkTXBAJMu2qaMWPrS7AANYqdq6vcBcBUdJCVVFceUvJFjaPdGZ2y9WACViL4L/0))##ggssrxfy", "sh(multi(2,[00000000/111'/222]xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHW ...
+    CheckUnparsable("sh(multi(2,[00000000/111'/222]xprvA1RpRA33e1JQ7ifknakTFpgNXPmW2YvmhqLQYMmrj4xJXXWYpDPS3xz7iAxn8L39njGVyuoseXzU6rcxFLJ8HFsTjSyQbLYnMpCqE2VbFWc,xprv9uPDJpEQgRQfDcW7BkF7eTya6RPxXeJCqCJGHuCJ4GiRVLzkTXBAJMu2qaMWPrS7AANYqdq6vcBcBUdJCVVFceUvJFjaPdGZ2y9WACViL4L/0))#ggssrxfy", "sh(multi(2,[00000000/111'/222]xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWk ...
+    CheckUnparsable("sh(multi(2,[00000000/111'/222]xprvA1RpRA33e1JQ7ifknakTFpgNXPmW2YvmhqLQYMmrj4xJXXWYpDPS3xz7iAxn8L39njGVyuoseXzU6rcxFLJ8HFsTjSyQbLYnMpCqE2VbFWc,xprv9uPDJpEQgRQfDcW7BkF7eTya6RPxXeJCqCJGHuCJ4GiRVLzkTXBAJMu2qaMWPrS7AANYqdq6vcBcBUdJCVVFceUvJFjaPdGZ2y9WACViL4L/0))##ggssrxfy", "sh(multi(2,[00000000/111'/222]xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHW ...
```
Why? These changes seem superfluous.

Sjors replied:
Indeed, dropping these and a few others.

darosior reviewed (commented):
I think you missed a couple places to switch from `'` to `h`. Also a question regarding `h` in normalized private descriptors.

Nit: if you retouch i think it would be nice to also update the fuzzing target `parse_hd_keypath` to call `FormatHDKeypath` with `apostrophe = false` since the following `WriteHDKeypath` would always call it with `true` anyways:
```patch
diff --git a/src/test/fuzz/parse_hd_keypath.cpp b/src/test/fuzz/parse_hd_keypath.cpp
index 411b70230a..d52f08c28a 100644
--- a/src/test/fuzz/parse_hd_keypath.cpp
+++ b/src/test/fuzz/parse_hd_keypath.cpp
@@ -18,6 +18,6 @@ FUZZ_TARGET(parse_hd_keypath)
 
     FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
     const std::vector<uint32_t> random_keypath = ConsumeRandomLengthIntegralVector<uint32_t>(fuzzed_data_provider);
-    (void)FormatHDKeypath(random_keypath);
+    (void)FormatHDKeypath(random_keypath, false);
     (void)WriteHDKeypath(random_keypath);
 }
```

Sjors commented:
Will rebase and address @darosior's comments soon(tm).

Sjors commented:
Rebased, added the suggested fuzzer coverage and fixed some `'` that I missed.

> What was the rationale for only normalizing the public-key version with h?

I was thinking along the lines that normalization (moving the xpub to a logical position, converting  `'` -> `h`) was an opt-in thing that we only do with the public-key version. However it's pretty much the default in practice, e.g. `listdescriptors` normalizes. We might as well do it for the private key version. The private key version isn't used for checksums, nor as the wallet identifier, so this change should be harmless. However this requires even more changes to the tests, so I'd rather not do this now.

Sjors commented:
Oops, didn't rebase far enough to get past the merge conflict...

achow101 commented on src/util/bip32.h:
```diff
@@ -14,6 +14,6 @@
 /** Write HD keypaths as strings */
 std::string WriteHDKeypath(const std::vector<uint32_t>& keypath);
-std::string FormatHDKeypath(const std::vector<uint32_t>& path);
+std::string FormatHDKeypath(const std::vector<uint32_t>& path, bool apostrophe = true);
```
Could this default to `false` so that we prefer `h`'s everywhere? In my own testing, this only caused a few functional tests to fail, the unit tests seem fine.

Sjors replied:
Done

Sjors replied:
Missed a few spots for legacy, pushed again.

achow101 commented:
reACK 4677f1923c9509025de0e77892638d5013f044bf

darosior commented on src/test/fuzz/parse_hd_keypath.cpp:
```diff
@@ -18,6 +18,6 @@ FUZZ_TARGET(parse_hd_keypath)
     FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
     const std::vector<uint32_t> random_keypath = ConsumeRandomLengthIntegralVector<uint32_t>(fuzzed_data_provider);
-    (void)FormatHDKeypath(random_keypath);
+    (void)FormatHDKeypath(random_keypath, false);
```
Since you changed the default for `apostrophe` in `FormatHDKeypath` to `false` this should be set to `true` here (as the below `WriteHDKeypaht` will already call it with `false`).

darosior commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1774,7 +1774,7 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
         KeyOriginInfo info;
         bool has_info = GetKeyOrigin(keyid, info);
-        std::string origin_str = has_info ? "[" + HexStr(info.fingerprint) + FormatHDKeypath(info.path) + "]" : "";
+        std::string origin_str = has_info ? "[" + HexStr(info.fingerprint) + FormatHDKeypath(info.path, false) + "]" : "";
```
nit: `false` is already the default for `FormatHDKeypath`.

Sjors replied:
Dropped.

darosior commented on test/functional/wallet_descriptor.py:
```diff
@@ -60,43 +60,43 @@ def run_test(self):
         addr_info = self.nodes[0].getaddressinfo(addr)
         assert addr_info['desc'].startswith('pkh(')
-        assert_equal(addr_info['hdkeypath'], 'm/44\'/1\'/0\'/0/0')
+        assert_equal(addr_info['hdkeypath'], 'm/44h/1h/0h/0/0')
```
This is an API break? (Albeit a small one)

Sjors replied:
I suppose it would be better if legacy wallet RPC calls don't change behavior. I'll look into this (and the other comments above).

Sjors replied:
The bigger break was actually in the wallet dump format. I've added handling to keep `hdkeypath` the same for legacy wallets.

darosior reviewed (commented):
Both the OP and the commit message state:
> The new apostrophe argument in FormatHDKeypath (bip32.h) ensures that the hdkeypath field in getaddressinfo is not impacted by this change.

But the commit then goes on to change it (because this code uses `WriteHDKeypath` which calls `FormatHDKeypath` with the default argument for `apostrophe` which you changed to `false` in the last push).

Sjors commented on doc/release-notes-26076.md:
```diff
@@ -0,0 +1,11 @@
+  hardened derivation. This does not apply when using the `private` parameter, which
+  matches the marker used when descriptor was generated or imported. Newly created
+  wallets use `h`, but previously created descriptor wallets still use `'`.
+  This change makes it easier to handle descriptor strings manually. E.g. an RPC call that
+  takes an array of descriptors can now use '["desc": ".../0h/..."]'. Either `h`
```
I modified the text a bit, and also updated the PR description.

Sjors committed:
Switch hardened derivation marker to h in descriptors

This makes it easier to handle descriptor strings manually. E.g. an RPC call that takes an array of descriptors can now use '["desc": ".../0h/..."]'.

Both markers can still be parsed. The default for new descriptors is changed to h. In normalized form h is also used. For private keys the chosen marker is preserved in a round trip.

The hdkeypath field in getaddressinfo is also impacted by this change.

achow101 commented:
re-ACK bd13dc2f46ea10302a928fcf0f53b7aed77ad260

darosior commented on doc/release-notes-26076.md:
```diff
@@ -0,0 +1,12 @@
+RPC
+---
+
+- The `listdescriptors` RPC now shows `h` rather than apostrophe (`'`) to indicate
```
`decodepsbt` too.

darosior reviewed (commented):
light ACK bd13dc2f46ea10302a928fcf0f53b7aed77ad260. Code looks correct to me but i find the OP and release note misleading (stated purpose is already possible and not what this PR implements) and i'm a bit confused about the choice of where to make breaking changes:
- Why not honour what was parsed for normalized serialization?
- Why keep serializing `hdkeypath`'s hardened paths with apostrophe for legacy wallet but not descriptor wallets or elsewhere (like in `decodepsbt`)?

Just to be clear: no strong opinion, and i don't mean to be blocking this. I think it's safe. Just i don't get the rationale for where to (slightly) break the API or not.

Sjors commented on src/wallet/rpc/addresses.cpp:
```diff
@@ -607,7 +607,9 @@ RPCHelpMan getaddressinfo()
             ret.pushKV("timestamp", meta->nCreateTime);
             if (meta->has_key_origin) {
-                ret.pushKV("hdkeypath", WriteHDKeypath(meta->key_origin.path));
+                // In legacy wallets hdkeypath has always used an apostrophe for
+                // hardened derivation. Perhaps some external tool depends on that.
+                ret.pushKV("hdkeypath", WriteHDKeypath(meta->key_origin.path, /*apostrophe=*/!desc_spk_man));
```
Descriptors have always required parsers to handle both `h` and `'`, and I'm not aware of any that don't. For old BIP32 derivation parsing software I'm less sure about that.

Sjors committed:
doc: clarify PR 26076 release note

darosior reviewed (commented):
re-ACK fe49f06c0e91b96feb8d8f1bd478c3173f14782c

achow101 commented:
ACK fe49f06c0e91b96feb8d8f1bd478c3173f14782c

Xekyo commented on doc/release-notes-26076.md:
```diff
@@ -0,0 +1,13 @@
+  With this change `listdescriptors` will use `h`, so you can copy-paste the result,
+  without having to add escape characters or switch `'` to 'h' manually.
+  Note that this changes the descriptor checksum.
+  For legacy wallets the `hdkeypath` field in `getaddressinfo` is unchanged,
+  nor is the serialization format of wallet dumps. (#26076)
```
The “nor” here seems misplaced:

```suggestion
  For legacy wallets the `hdkeypath` field in `getaddressinfo` 
  and the serialization format of wallet dumps remain unchanged. (#26076)
```
